File: dfs/provinces/provinces.cpp
#include <vector>
#include <unordered_set>
using namespace std;

class Provinces // LeetCode Q.547.
{
private:
    vector<vector<int>> graph;
    unordered_set<int> visitedCities;
    int totalCapitals = 0;

    void dfsProvinces(int city)
    {
        visitedCities.insert(city);
        for (int nextCity = 0; nextCity < graph[city].size(); nextCity++)
        {
            if (graph[city][nextCity] == 1)
            {
                if (visitedCities.find(nextCity) == visitedCities.end())
                    dfsProvinces(nextCity);
            }
        }
    }

public:
    int countProvinces(vector<vector<int>> &connections)
    {
        graph = connections;
        for (int city_1 = 0; city_1 < graph.size(); city_1++)
        {
            if (visitedCities.find(city_1) == visitedCities.end())
            {
                totalCapitals++;
                dfsProvinces(city_1);
            }
        }

        return totalCapitals;
    }
};

File: dfs/provinces/provinces.py

class Provinces:  # LeetCode Q.547.
    def count_provinces(self, connections: list[list[int]]) -> int:
        self.graph: list[list[int]] = connections
        self.visited_cities: set[int] = set()
        total_capitals = 0

        for city_1 in range(len(self.graph)):
            if city_1 not in self.visited_cities:
                total_capitals +=1
                self._dfs_provinces(city_1)

        return total_capitals

    def _dfs_provinces(self, city: int) -> None:
        self.visited_cities.add(city)
        for next_city in range(len(self.graph[city])):
            if self.graph[city][next_city] == 1:
                if next_city not in self.visited_cities:
                    self._dfs_provinces(next_city)
